server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -514,4 +506,437 @@ boolean containsBrokenAnalysis(String field) {
+     */
+    private static class EmptySnapshot extends AbstractSnapshot {
+        EmptySnapshot(MapperService mapperService) {
+            super(mapperService);
+        }
+
+        @Override
+        public MappedFieldType fieldType(String fullName) {
+            if (fullName.equals(TypeFieldType.NAME)) {
+                return new TypeFieldType("_doc");
+            }
+            return null;
+        }
+
+        @Override
+        public boolean hasNested() {
+            return false;
+        }
+
+        @Override
+        public ObjectMapper getObjectMapper(String name) {
+            return null;
+        }
+
+        @Override
+        public Set<String> simpleMatchToFullName(String pattern) {
+            return Regex.isSimpleMatchPattern(pattern) ? emptySet() : singleton(pattern);
+        }
+
+        @Override
+        public Set<String> sourcePath(String fullName) {
+            return emptySet();
+        }
+
+        @Override
+        public DocumentMapperForType documentMapperWithAutoCreate() {
+            DocumentMapper auto = mapperService.parse(SINGLE_MAPPING_NAME, null);
+            return new DocumentMapperForType(auto, auto.mapping());
+        }
+
+        @Override
+        public Iterable<MappedFieldType> getEagerGlobalOrdinalsFields() {
+            return emptySet();
+        }
+
+        @Override
+        public FieldNameAnalyzer indexAnalyzer() {
+            return null;
+        }
+
+        @Override
+        public boolean containsBrokenAnalysis(String field) {
+            return false;
+        }
+
+        @Override
+        public DocumentMapper documentMapper() {
+            return null;
+        }
+
+        @Override
+        public long version() {
+            return 0;
+        }
+
+        @Override
+        public ParsedDocument parseDocument(SourceToParse source) throws MapperParsingException {
+            return null;
+        }
+
+        @Override
+        Map<Class<? extends MetadataFieldMapper>, MetadataFieldMapper> getMetadataMappers() {
+            Map<String, MetadataFieldMapper.TypeParser> metadataMapperParsers = mapperService.mapperRegistry.getMetadataMapperParsers(
+                mapperService.indexSettings.getIndexVersionCreated()
+            );
+            Map<Class<? extends MetadataFieldMapper>, MetadataFieldMapper> metadataMappers = new LinkedHashMap<>();
+            for (MetadataFieldMapper.TypeParser parser : metadataMapperParsers.values()) {
+                MetadataFieldMapper metadataFieldMapper = parser.getDefault(parserContext());
+                metadataMappers.put(metadataFieldMapper.getClass(), metadataFieldMapper);
+            }
+            return metadataMappers;
+        }
+
+        @Override
+        MappedSnapshot merge(DocumentMapper mapper, MergeReason reason) {
+            return new MappedSnapshot(mapperService, mapper, 1);
+        }
+
+        @Override
+        protected String updateOperationName() {
```

Comment:
this is some odd looking method, does it belong to the snapshot?

Reply:
Its the name of the operation that we log when you merge something into the snapshot. Its `protected` so its part of the internal interface for snapshot and doesn't escape outside of `MapperService`. I kind of like that we can have the internal interface and the external one. Stuff like this that is convenient to push over to the snapshot we can push without it leaking.